src/main/java/com/google/cloud/partners/pubsub/kafka/SubscriberService.java

```diff
@@ -0,0 +1,460 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Implementation of <a
+ * href="https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#subscriber"
+ * target="_blank"> Cloud Pub/Sub Publisher API.</a>
+ */
+@Singleton
+class SubscriberService extends SubscriberImplBase {
+
+  private static final AtomicInteger STREAMING_PULL_ID = new AtomicInteger();
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+
+  private final Map<String, SubscriptionManager> subscriptions;
+  private final StatisticsManager statisticsManager;
+  private final ConfigurationManager configurationManager;
+  private final SubscriptionManagerFactory subscriptionManagerFactory;
+
+  @Inject
+  SubscriberService(
+      ConfigurationManager configurationManager,
+      SubscriptionManagerFactory subscriptionManagerFactory,
+      StatisticsManager statisticsManager) {
+    this.configurationManager = configurationManager;
+    this.statisticsManager = statisticsManager;
+    this.subscriptionManagerFactory = subscriptionManagerFactory;
+
+    subscriptions =
+        configurationManager
+            .getProjects()
+            .stream()
+            .flatMap(p -> configurationManager.getSubscriptions(p).stream())
+            .collect(
+                Collectors.toConcurrentMap(
+                    Subscription::getName,
+                    subscription -> {
+                      SubscriptionManager sm = subscriptionManagerFactory.create(subscription);
+                      logger.atFine().log(
+                          "Starting Subscription Manager for %s", subscription.getName());
+                      sm.startAsync().awaitRunning();
+                      logger.atInfo().log(
+                          "Started Subscription Manager for %s", subscription.getName());
+                      return sm;
+                    }));
+  }
+
+  /**
+   * Shutdown hook halts the scheduled executor service preventing new commit tasks from being
+   * submitted, executes commits for each {@link SubscriptionManager}, and then closes all
+   * KafkaConsumers.
+   */
+  public void shutdown() {
+    subscriptions
+        .values()
+        .forEach(
+            sm -> {
+              logger.atFine().log(
+                  "Stopping Subscription Manager for %s", sm.getSubscription().getName());
+              sm.stopAsync().awaitTerminated();
+              logger.atFine().log(
+                  "Stopped Subscription Manager for %s", sm.getSubscription().getName());
+            });
+  }
+
+  @Override
+  public void createSubscription(
+      Subscription request, StreamObserver<Subscription> responseObserver) {
+    try {
+      logger.atFine().log("Creating Subscription %s", request);
+      configurationManager.createSubscription(request);
```

Comment:
I found out that subscriptions created at runtime through the API are unable to receive messages. Here's how to reproduce:
1. start the emulator
2. create a new subscription to a topic
3. publish any message to the topic
4. execute a pull on the subscription at 2
5. no messages are received

This is because on 'createSubscription` we pass the `request` subscription to the `subscriptionManagerFactory` instead of the one in the `configurationManager` map. The main difference is that the former lacks the `KAFKA_TOPIC` label introduced by the `configurationManager`.

The absence of the label causes the kafka consumer to fail because it defaults to the `/projects/{project}/topics/{topic}` PubSub topic name which is invalid in Kafka.

This bug was very hard to pin down because the `SubscriptionManager` scheduled service fails silently without errors. It becomes evident if when calling `startUp` explicitly in the `SubscriptionManager` constructor. At this point, every `createSubscription` generates a similar error:
```
[Thu Feb 28 16:59:14 GMT 2019] SEVERE: io.grpc.internal.SerializingExecutor run - Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed@33de6819 
org.apache.kafka.common.errors.InvalidTopicException: Topic 'projects/default/topics/t3' is invalid
```
Additionally, I can confirm that the following patch fixes the issue:
```
diff --git a/src/main/java/com/google/cloud/partners/pubsub/kafka/SubscriberService.java b/src/main/java/com/google/cloud/partners/pubsub/kafka/SubscriberService.java
index e678297..99e9ece 100644
--- a/src/main/java/com/google/cloud/partners/pubsub/kafka/SubscriberService.java
+++ b/src/main/java/com/google/cloud/partners/pubsub/kafka/SubscriberService.java
@@ -117,11 +117,13 @@ class SubscriberService extends SubscriberImplBase {
   @Override
   public void createSubscription(
       Subscription request, StreamObserver<Subscription> responseObserver) {
+    Subscription configuredSubscription;
     try {
       logger.atFine().log("Creating Subscription %s", request);
       configurationManager.createSubscription(request);
-      subscriptions.put(request.getName(), subscriptionManagerFactory.create(request));
-      statisticsManager.addSubscriberInformation(request);
+      configuredSubscription = configurationManager.getSubscriptionByName(request.getName()).get();
+      subscriptions.put(request.getName(), subscriptionManagerFactory.create(configuredSubscription));
+      statisticsManager.addSubscriberInformation(configuredSubscription);
       responseObserver.onNext(request);
       responseObserver.onCompleted();
     } catch (ConfigurationAlreadyExistsException e) {
```

Reply:
My sincere apologies for this but thank you for the diligent debugging efforts. I should have made the ConfigurationManager methods return the persisted objects since they do mutate them. I'm introducing that change and will have a patch shortly.

Reply:
I believe that `SubscriptionManager` object are now created with the right `Subscription` object however it seems they're not scheduled.

Sending messages to newly created subscriptions are not delivered unless I explicitly call `startUp` in the `SubscriptionManager` constructor.

I am not sure I can spot any difference with the ones created from the config file. Any idea?

Reply:
Sorry that was another omission on my part. When deleting a Subscription I was making sure that we stopped it's manager `     subscriptions.get(request.getSubscription()).stopAsync().awaitTerminated();` but was not doing the inverse when creating a new Subscription.

I've added that code and tests to confirm its behavior.

Reply:
I can confirm that my test pass now! Thank you!